### PR TITLE
IR changes for storage accesses

### PR DESCRIPTION
**Pull request by mohammadfawaz**
I've decided to split the storage access work in multiple PRs. 

I do have most of the work done for the full flow, but the change is too big to review effectively. This first PR handles IR and IR -> ASM.

### Changes in the IR
* Modified the `get_ptr` instruction to also take a pointer type and an offset: `get_ptr <base_ptr>, <ptr_ty>, <offset>`. The instruction now converts`base_ptr` to `ptr_ty` and then offsets the result by `offset`. The unit of `offset` here is basically the size of `ptr_ty`. So, if `ptr_ty` is `ptr u64` and `offset` is `5`, then we basically have to add `5*sizeof(ptr_ty)` to `base_ptr`.
* Removed `ptr_cast` for now as its functionality can now be accomplished using `get_ptr`.
* Replaced `state_load` and `state_store` with 4 instructions:
  * `state_load_word <key>`: returns a word (`u64`) from the storage location `key`. `key` has to be a `b256`.
  * `state_load_quad_word <val>, <key>`: reads 4 words (`b256`) from storage location `key` into `val` where `val` is a pointer to a `b256`.
  * `state_store_word <val>, <key>`: stores `val` to storage location `key`. `val` has to be a `u64`.
  * `state_store_quad_word <val>, <key>`: stores 4 words pointed to by `val` to storage location `key`.
* Modified the parser to handle contracts and function selectors

### Changes in IR -> ASM
* Basically supporting the 4 new IR instructions above and the new syntax for `get_ptr`. 

### Notes
* The IR generator is expected to break each storage access into a list of `state_load_*`/`state_store_*`.
* The ASM generator will need some additional updates to support the new `get_ptr` arguments, especially `load` and `store`, but for now, the new arguments are only exercised for `state_load_*` and `state_store_*`.

### CODE REVIEW

File: sway-ir/src/parser.rs [hunk 1]
```diff
@@ -52,9 +65,24 @@ mod ir_builder {
...
+                    let mut cur_byte: u8 = 0;
+                    for (idx, ch) in s.chars().enumerate() {
+                        cur_byte = (cur_byte << 4) | ch.to_digit(16).unwrap() as u8;
+                        if idx % 2 == 1 {
+                            bytes[idx / 2] = cur_byte;
+                            cur_byte = 0;
+                        }
+                    }
+                    bytes
+                }
```
Comment (otrho) [inline sway-ir/src/parser.rs:84]: I was looking at this thinking it could be done as a one liner, but the best I could come up with is:
```rust
let cnv = |c: char| c.to_digit(16).unwrap() as u8;
s.chunks(2).map(|ch| (cnv(ch[0]) << 4) | cnv(ch[1])).collect::<Vec<_>>().try_into().unwrap()
```
...which I'm not convinced is better. 🙂

File: sway-ir/src/printer.rs [hunk 2]
```diff
@@ -476,14 +475,26 @@ fn instruction_to_doc<'a>(
                     md_namer.meta_as_string(context, span_md_idx, true),
                 )))
             }
-            Instruction::StateLoad { load_val, key } => Doc::text_line(format!(
-                "state_load ptr {}, key {}{}",
+            Instruction::StateLoadQuadWord { load_val, key } => Doc::text_line(format!(
+                "state_load_quad_word ptr {}, ptr {}{}",
                 namer.name(context, load_val),
```
Comment (otrho) [inline sway-ir/src/printer.rs:480]: You've switched the `key` keyword to be another `ptr`, which I think might be confusing.  When reading the IR it isn't going to be obvious which argument is which.  Maybe `state_load_quad_word ptr {}, key ptr {}{}` is a compromise?

**Review by otrho — CHANGES_REQUESTED** [review]
Looks great. 🙂

File: sway-core/src/asm_generation/from_ir.rs [hunk 3]
```diff
@@ -1196,6 +1238,255 @@ impl<'ir> AsmBuilder<'ir> {
...
+                            key_offset * 8,
+                            instr_val.get_span(self.context),
+                        );
+
+                        self.bytecode.push(Op {
+                            opcode: Either::Left(VirtualOp::SWWQ(store_val_reg, key_reg)),
+                            comment: "quad state store value".into(),
+                            owning_span: instr_val.get_span(self.context),
+                        });
+                    }
```
[the remark this answers is not in the document]
Comment (mohammadfawaz) [reply sway-core/src/asm_generation/from_ir.rs]: Done. I wanted to merge the other two functions as well but they're different enough that merging may cause confusion, so I didn't.

File: sway-ir/src/parser.rs (hunk 1 above)
Comment (mohammadfawaz) [reply sway-ir/src/parser.rs:77]: This code is mostly a duplicate of another that you wrote to parser `b256` values. I ended up merging the two in one helper function.

File: sway-ir/src/printer.rs (hunk 2 above)
Comment (mohammadfawaz) [reply sway-ir/src/printer.rs:480]: I like having `ptr` there to indicate that this is indeed a pointer to a b256, so I now have `key ptr {}` as per your recommendation.

**Review by otrho — APPROVED** [review]
Cool.
